sky/provision/paperspace/instance.py

```diff
@@ -0,0 +1,264 @@
+    ]
+    client = PaperspaceCloudClient()
+    while True:
+        instances = _filter_instances(cluster_name_on_cloud, pending_status)
+        if not instances:
+            break
+        for instance_id, instance_meta in instances.items():
+            if instance_meta['state'] == 'off':
+                try:
+                    client.start(instance_id=instance_id)
+                except PaperspaceCloudError as e:
+                    if 'This machine is currently starting.' in str(e):
+                        continue
+                    raise e
+        logger.info(f'Waiting for {len(instances)} instances to be ready.')
+        time.sleep(POLL_INTERVAL)
+    exist_instances = _filter_instances(cluster_name_on_cloud, ['ready'])
+    head_instance_id = _get_head_instance_id(exist_instances)
+
+    to_start_count = config.count - len(exist_instances)
+    if to_start_count < 0:
+        raise RuntimeError(
+            f'Cluster {cluster_name_on_cloud} already has '
+            f'{len(exist_instances)} nodes, but {config.count} are required.')
+    if to_start_count == 0:
+        if head_instance_id is None:
+            raise RuntimeError(
+                f'Cluster {cluster_name_on_cloud} has no head node.')
+        logger.info(f'Cluster {cluster_name_on_cloud} already has '
+                    f'{len(exist_instances)} nodes, no need to start more.')
+        return common.ProvisionRecord(
+            provider_name='paperspace',
+            cluster_name=cluster_name_on_cloud,
+            region=region,
+            zone=None,
+            head_instance_id=head_instance_id,
+            resumed_instance_ids=[],
+            created_instance_ids=[],
+        )
+
+    created_instance_ids = []
+    for _ in range(to_start_count):
+        node_type = 'head' if head_instance_id is None else 'worker'
+        try:
+            instance_id = client.launch(
+                name=f'{cluster_name_on_cloud}-{node_type}',
+                instance_type=config.node_config['InstanceType'],
+                network_id=config.node_config['NetworkId'],
+                region=region,
+                disk_size=config.node_config['DiskSize'],
+            )['data']['id']
+        except Exception as e:  # pylint: disable=broad-except
+            logger.warning(f'run_instances error: {e}')
+            raise e
+        logger.info(f'Launched instance {instance_id}.')
+        created_instance_ids.append(instance_id)
+        if head_instance_id is None:
+            head_instance_id = instance_id
+
+    # Wait for instances to be ready.
+    while True:
+        instances = _filter_instances(cluster_name_on_cloud, ['ready'])
+        logger.info('Waiting for instances to be ready: '
+                    f'({len(instances)}/{config.count}).')
+        if len(instances) == config.count:
+            break
+
+        time.sleep(POLL_INTERVAL)
+    assert head_instance_id is not None, 'head_instance_id should not be None'
```

Comment:
Question: do we have a state on paperspace that indicates an instance is stopping or terminating? We should probably wait those VMs to be in a terminal state before we restart them.

We should wait for the instances being started as well, before we filter the `exist_instances` by ready status. Also, we can fail early before starting the stopped instances if the number of existing instances is more than the instances requested.

```suggestion
    pending_status = [
        'starting', 'restarting', 'upgrading', 'provisioning'
    ]
    newly_started_instances = _filter_instances(cluster_name_on_cloud, pending_status + ['off'])
    client = PaperspaceCloudClient()
    while True:
        instances = _filter_instances(cluster_name_on_cloud, pending_status)
        if not instances:
            break
        logger.info(f'Waiting for {len(instances)} instances to be ready.')
        time.sleep(POLL_INTERVAL)
    exist_instances = _filter_instances(cluster_name_on_cloud, status_filters=pending_status + ['ready', 'off'])
    if len(exist_instances) > config.count:
        raise RuntimeError(
            f'Cluster {cluster_name_on_cloud} already has '
            f'{len(exist_instances)} nodes, but {config.count} are required.')
       
    stopped_instances = _filter_instances(cluster_name_on_cloud, status_filters=['off'])
    for instance_id, instance_meta in stopped_instances.items():
        try:
            client.start(instance_id=instance_id)
        except PaperspaceCloudError as e:
            if 'This machine is currently starting.' in str(e):
                continue
            raise e
           
    to_start_count = config.count - len(exist_instances)
    created_instance_ids = []
    if to_start_count > 0:
        for _ in range(to_start_count):
            node_type = 'head' if head_instance_id is None else 'worker'
            try:
                instance_id = client.launch(
                    name=f'{cluster_name_on_cloud}-{node_type}',
                    instance_type=config.node_config['InstanceType'],
                    network_id=config.node_config['NetworkId'],
                    region=region,
                    disk_size=config.node_config['DiskSize'],
                )['data']['id']
            except Exception as e:  # pylint: disable=broad-except
                logger.warning(f'run_instances error: {e}')
                raise e
            logger.info(f'Launched instance {instance_id}.')
            created_instance_ids.append(instance_id)
            if head_instance_id is None:
                head_instance_id = instance_id
    # Wait for instances to be ready.
    while True:
        instances = _filter_instances(cluster_name_on_cloud, ['ready'])
        logger.info('Waiting for instances to be ready: '
                    f'({len(instances)}/{config.count}).')
        if len(instances) == config.count:
            break
```